module/zfs/sa.c

```diff
@@ -1496,6 +1500,185 @@ sa_lookup_uio(sa_handle_t *hdl, sa_attr_type_t attr, uio_t *uio)
 	mutex_exit(&hdl->sa_lock);
 	return (error);
 }
+
+static int
+sa_lookup_locked(sa_handle_t *hdl, sa_attr_type_t attr, void *buf,
+    uint32_t buflen)
+{
+	int error;
+	sa_bulk_attr_t bulk;
+
+	VERIFY3U(buflen, <=, SA_ATTR_MAX_LEN);
+
+	bulk.sa_attr = attr;
+	bulk.sa_data = buf;
+	bulk.sa_length = buflen;
+	bulk.sa_data_func = NULL;
+
+	ASSERT(hdl);
+	error = sa_lookup_impl(hdl, &bulk, 1);
+	return (error);
+}
+
+int
+sa_add_projid(sa_handle_t *hdl, dmu_tx_t *tx, uint64_t projid)
```

Comment:
In theory, it is not impossible to implement it via the zfs_sa_upgrade() path. But we need more circuitous changes, for example:
1) New interface (param or flag) to make the zfs_sa_upgrade() can be triggered inside sa_bulk_update_impl(). Looks not clean.

2) Make zfs_sa_upgrade() to handle the given object by force, not skip because of z_acl_cached or similar. That may affect other use cases.
>         /*
>         * No upgrade if ACL isn't cached
>         * since we won't know which locks are held
>         * and ready the ACL would require special "locked"
>         * interfaces that would be messy
>         */         
>        if (zp->z_acl_cached == NULL || S_ISLNK(ZTOI(zp)->i_mode))
>                return;

3) Current zfs_sa_upgrade() does not care whether the real upgrade is done or skipped because it can be upgraded at a later time. But for set_projid case, we have to that at once.
>         /*
>         * If the z_lock is held and we aren't the owner
>         * the just return since we don't want to deadlock
>         * trying to update the status of z_is_sa.  This
>         * file can then be upgraded at a later time.
>         *
>         * Otherwise, we know we are doing the
>         * sa_update() that caused us to enter this function.
>         */
>        if (MUTEX_NOT_HELD(&zp->z_lock)) {
>                if (mutex_tryenter(&zp->z_lock) == 0)
>                        return;
>                else
>                        drop_lock = B_TRUE;
>        }


More important, I hope to export the "sa_add_projid()", that may be used by up layer callers, such as Lustre.

So currently, I prefer to make the sa_add_projid() to be independent from zfs_sa_upgarde().